src/engagement_db_to_analysis/code_imputation_functions.py

```diff
@@ -222,6 +222,42 @@ def _impute_ws_coding_errors(user, messages_traced_data, analysis_dataset_config
     log.info(f"Imputed {imputed_labels} {Codes.CODING_ERROR} labels for WS codes")
 
 
+def _impute_nc_for_empty_messages(user, messages_traced_data, analysis_dataset_configs, ws_correct_dataset_code_scheme):
+    """
+    Imputes Codes.NOT_CODED for messages whose text property is either None or "".
+
+    :param user: Identifier of user running the pipeline.
+    :type user: str
+    :param messages_traced_data: Messages TracedData objects to impute age_category.
+    :type messages_traced_data: list of TracedData
+    :param analysis_dataset_configs: Analysis dataset configuration in pipeline configuration module.
+    :type analysis_dataset_configs: pipeline_config.analysis_configs.dataset_configurations
+    :param ws_correct_dataset_code_scheme: WS - Correct Dataset code scheme.
+    :type ws_correct_dataset_code_scheme: core_data_modules.data_models.CodeScheme
+    """
+    messages_with_nc_imputed = 0
+    for message_td in messages_traced_data:
+        message = Message.from_dict(dict(message_td))
+
+        if not (message.text is None or message.text == ""):
```

Comment:
Have we encountered this from Radpidpro before message.text is None ?

Reply:
Yes - I checked old data from a year ago and ssf elections and saw this in both cases, but it looks like it's when the text received from Shaqodoon is "". I think we should ask (i) Shaqodoon what causes an "" to be sent, and (ii) Rapid Pro if auto-conversion to null is the right thing to do. Perhaps we can discuss tomorrow

Reply:
I've checked the logs more carefully - the message from Shaqodoon contains 'text=', so I can understand why Rapid Pro interprets this as null. (It converts to "" in its acknowledgement reply to Shaqodoon, but I don't think that's a problem)